cdap-app-fabric/src/main/java/io/cdap/cdap/support/services/SupportBundleService.java

```diff
@@ -0,0 +1,194 @@
+      } else {
+        namespaceList.add(namespaceId);
+      }
+      // Puts all the files under the uuid path
+      File baseDirectory = new File(localDir);
+      DirUtils.mkdirs(baseDirectory);
+      int fileCount = DirUtils.list(baseDirectory).size();
+
+      // We want to keep consistent number of bundle to provide to customer
+      if (fileCount >= folderMaxNumber) {
+        File oldFilesDirectory = getOldestFolder(baseDirectory);
+        deleteOldFolders(oldFilesDirectory);
+      }
+      DirUtils.mkdirs(basePath);
+      //TODO Execute tasks in parallel
+    } catch (Exception e) {
+      LOG.error("Failed to generate support bundle ", e);
+      supportBundleStatus.setFinishTimestamp(System.currentTimeMillis());
+      supportBundleStatus.setStatus(CollectionState.FAILED);
+      supportBundleStatus.setStatusDetails(e.getMessage());
+      addToStatus(supportBundleStatus, basePath.getPath());
+      throw new Exception("Failed to generate support bundle ", e);
+    }
+    return uuid;
+  }
+
+  /**
+   * Gets oldest folder from the root directory
+   */
+  private File getOldestFolder(File baseDirectory) throws RuntimeException {
+    File[] uuidFiles =
+      baseDirectory.listFiles((dir, name) -> !name.startsWith(".") && !dir.isHidden() && dir.isDirectory());
+    return Collections.min(Arrays.asList(uuidFiles), Comparator.<File, Boolean>comparing(f1 -> {
+      try {
+        return getSingleBundleJson(f1).getStatus() != CollectionState.FAILED;
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to get file status ", e);
+      }
+    }).thenComparing(File::lastModified));
+  }
+
+  /**
+   * Deletes old folders after certain number of folders exist
+   */
+  private void deleteOldFolders(@Nullable File oldFilesDirectory) {
+    try {
+      DirUtils.deleteDirectoryContents(oldFilesDirectory);
+    } catch (IOException e) {
+      LOG.error("Failed to clean up directory {}", oldFilesDirectory, e);
+    }
+  }
+
+  public void close() {
+    this.executorService.shutdown();
+  }
+
+  /**
+   * Update status file
+   */
+  private void addToStatus(SupportBundleStatus supportBundleStatus, String basePath) throws Exception {
+    try (FileWriter statusFile = new FileWriter(new File(basePath, SupportBundleFileNames.statusFileName))) {
+      GSON.toJson(supportBundleStatus, statusFile);
+      statusFile.flush();
+    } catch (Exception e) {
+      LOG.error("Can not update status file ", e);
+      throw new Exception("Can not update status file ", e);
+    }
+  }
+
+  /**
+   * Gets single support bundle status with uuid
+   */
+  public SupportBundleStatus getSingleBundleJson(File uuidFile) throws Exception {
+    File statusFile = new File(uuidFile, SupportBundleFileNames.statusFileName);
+    if (!statusFile.exists()) {
+      LOG.error("Can not find this status file {} ", uuidFile.getName());
+      throw new Exception("Can not find this status file");
```

Comment:
Don't throw `Exception`. You can introduce a new one or throw `IllegarArgumentException`. Using runtime here will save you some rethrowing